### ISSUE TITLE
Smart Search and Recommendation System Principles Algorithm and Application

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9297]
Smart Search and Recommendation System Principles, Algorithm and Application<br>
<br>
https://ift.tt/tNkKYVR<br>
<br>
贪钱算法还我头发<br>
<br>

<div><div><div><div><article><p>《智能搜索和推荐系统原理、算法和应用》读书笔记，主要记录了搜索推荐中常用的算法，具体的算法流程需要单独记录笔记。书中代码<a rel="noopener" href="https://github.com/michaelliu03/Search-Recommend-InAction" target="_blank">https://github.com/michaelliu03/Search-Recommend-InAction</a>。</p>
<h1>搜索和推荐系统的基础</h1>
<h2>概率论统计与应用数学基础知识</h2>
<h3>概率论基础</h3>
<ol>
<li>
<p><strong>古典概率</strong>：可重复的实验次数N趋于无穷时，事件A发生的频率Q无限接近概率P。$$lim_{N\rightarrow \infty}Q(A)=P(A)$$</p>
</li>
<li>
<p><strong>条件概率</strong>：已知事件B发生的情况下事件A发生的概率。$$P(A|B)=\frac{P(AB)}{P(B)}$$</p>
</li>
<li>
<p><strong>全概率公式</strong>：将事件A分割成小事件，先求小事件的概率，相加求得事件A的概率$$P(A)=P\left(A\bigcap(\bigcup^n_{i=1}B_i)\right)=\sum^n_{i=1}P(AB_i)=\sum^n_{i=1}P(B_i)P(A|B_i)$$</p>
</li>
<li>
<p><strong>贝叶斯公式</strong>：在大事件A发生的条件下求分割的小事件$B_i$发生的概率。$$P(B_i|A)=\frac{P(B_i)P(A|B_i)}{P(A)}=\frac{P(B_i)P(A|B_i)}{\sum^n_{i=1}P(B_i)P(A|B_i)}$$</p>
<p>$$后验=\frac{先验·似然}{P(A)}$$</p>
</li>
<li>
<p><strong>基础的概率分布</strong>：</p>
<ul>
<li>0-1分布：$P(X=k)=p^k(1-p)^{1-k},k=0,1$</li>
<li>二项分布$B(n,p)$：$P(X=k)=C^k_np^k(1-p)^{n-k},k=0,1…n$</li>
<li>正态分布$N(\mu,\sigma^2)$：$\psi(x)=\frac{1}{\sqrt{2\pi}\sigma}e^{-\frac{(x-\mu)^2}{2\sigma^2}}$</li>
<li>柏松分布</li>
<li>均匀分布</li>
<li>指数分布</li>
<li>几何分布</li>
<li>超几何分布</li>
</ul>
</li>
</ol>
<figure><table><tr>
<td><pre><span>1</span><br><span>2</span><br><span>3</span><br><span>4</span><br><span>5</span><br><span>6</span><br><span>7</span><br><span>8</span><br><span>9</span><br><span>10</span><br><span>11</span><br><span>12</span><br><span>13</span><br><span>14</span><br><span>15</span><br><span>16</span><br><span>17</span><br><span>18</span><br><span>19</span><br><span>20</span><br><span>21</span><br><span>22</span><br><span>23</span><br><span>24</span><br><span>25</span><br><span>26</span><br><span>27</span><br><span>28</span><br><span>29</span><br><span>30</span><br><span>31</span><br><span>32</span><br><span>33</span><br><span>34</span><br><span>35</span><br><span>36</span><br><span>37</span><br><span>38</span><br></pre></td>
<td><pre><span><span>import</span> numpy <span>as</span> np</span><br><span><span>from</span> scipy <span>import</span> stats</span><br><span><span>import</span> matplotlib.pyplot <span>as</span> plt</span><br><span></span><br><span><span><span>def</span> <span>binom_pmf_test</span>():</span></span><br><span> <span>'''</span></span><br><span><span> 离散分布</span></span><br><span><span> 二项分布的例子：抛掷100次硬币，恰好两次正面朝上的概率是多少？</span></span><br><span><span> '''</span></span><br><span> n = <span>100</span> </span><br><span> p = <span>0.5</span> </span><br><span> k = np.arange(<span>0</span>, <span>100</span>) </span><br><span> binomial = stats.binom.pmf(k, n, p)</span><br><span> <span>print</span>(<span>sum</span>(binomial)) </span><br><span> plt.plot(k, binomial, <span>'o-'</span>)</span><br><span> plt.title(<span>'Binomial: n=%i , p=%.2f'</span> % (n, p), fontsize=<span>15</span>)</span><br><span> plt.xlabel(<span>'Number of successes'</span>)</span><br><span> plt.ylabel(<span>'Probability of success'</span>, fontsize=<span>15</span>)</span><br><span> plt.show()</span><br><span></span><br><span><span><span>def</span> <span>normal_distribution</span>():</span></span><br><span> <span>'''</span></span><br><span><span> 正态分布是一种连续分布，其函数可以在实线上的任何地方取值。</span></span><br><span><span> 正态分布由两个参数描述：分布的平均值μ和方差σ2 。</span></span><br><span><span> '''</span></span><br><span> mu = <span>0</span> </span><br><span> sigma = <span>1</span> </span><br><span> x = np.arange(-<span>10</span>, <span>10</span>, <span>0.1</span>)</span><br><span> y = stats.norm.pdf(x, <span>0</span>, <span>1</span>)</span><br><span> plt.plot(x, y)</span><br><span> plt.title(<span>'Normal: $\mu$=%.1f, $\sigma^2$=%.1f'</span> % (mu, sigma))</span><br><span> plt.xlabel(<span>'x'</span>)</span><br><span> plt.ylabel(<span>'Probability density'</span>, fontsize=<span>15</span>)</span><br><span> plt.show()</span><br><span></span><br><span><span>if</span> __name__ ==<span>"__main__"</span>:</span><br><span> binom_pmf_test() </span><br><span> normal_distribution() </span><br></pre></td>
</tr></table></figure><center>
<img alt="二项分布" align="center" src="https://xfliu1998.github.io/2021/12/03/4-Smart-Search-and-Recommendation-System-Principles-Algorithm-and-Application/binom.png"><p>二项分布</p>
<img alt="正态分布" align="center" src="https://xfliu1998.github.io/2021/12/03/4-Smart-Search-and-Recommendation-System-Principles-Algorithm-and-Application/normal.png"><p>正态分布</p>
</center>
<ul>
<li>期望$E(X)$</li>
<li>方差$D(X)$</li>
<li>标准差$\sqrt{D(X)}$</li>
<li>协方差$Cov(X,Y)=E[(X-E(X))(Y-E(Y))]$</li>
</ul>
<h3>线性代数基础</h3>
<ol>
<li>
<p>矩阵</p>
</li>
<li>
<p>向量</p>
</li>
<li>
<p>张量</p>
</li>
<li>
<p><strong>特征向量和特征值</strong>：矩阵A的特征向量指与A相乘后相当于对该向量进行缩放的非0向量x，$\lambda$是相应的特征值。$$Ax=\lambda x$$</p>
<p> $$x^TA=\lambda x^T$$</p>
<p> 假设A有n个线性无关的特征向量${v^{(1)},…,v^{(n)}}$，V是每一列是一个特征向量的矩阵，将A特征值分解：$$A=Vdiag(\lambda)V^{-1}$$</p>
<p> 不是每一个矩阵都能分解为特征值和特征向量，特征分解可能会涉及复数和非实数。每一个实对称矩阵都能分解为实特征向量和实特征值。$$A=Q\Lambda Q^{-1}$$</p>
<p> 其中Q是A的特征向量组成的正交矩阵（若 $AA^T=E$，n阶实矩阵A是正交矩阵），$\Lambda$是对角矩阵。</p>
</li>
<li>
<p><strong>奇异值分解</strong>：可以将非方阵分解为奇异向量和奇异值。$$A_{m\times n}=U_{m\times m}D_{m\times n}V^T_{n\times n}$$</p>
<p> U和V是正交阵，D是对角阵，对角阵上的元素称为矩阵A的奇异值，U的列向量为左奇异向量，V的列向量为右奇异向量。</p>
</li>
</ol>
<h3>机器学习基础</h3>
<ol>
<li>
<p>导数</p>
</li>
<li>
<p>梯度（矢量）</p>
</li>
<li>
<p><strong>最大似然估计</strong>：已知某个总体下的随机样本满足某种概率分布，概率分布的参数是未知的，经过反复试验，若某个参数使得样本出现概率最大，则这个参数值当作最大似然估计值。</p>
</li>
<li>
<p>随机过程与隐马尔可夫模型</p>
</li>
<li>
<p><strong>信息熵</strong>：熵是对不确定性和无序程度的预测。熵越大信息越混乱越不确定。</p>
<ul>
<li>熵：$H(X)=-\sum_{x\in R}p(x)log_2p(x)$，均匀概率分布时熵最大</li>
<li>联合熵：一对离散随机变量的不确定性$H(X,Y)=\sum_{x\in \Omega}\sum_{y\in \Psi}p(x,y)log_2p(x,y)$</li>
<li>条件熵：给定随机变量X条件下Y的条件熵$H(Y|x)=\sum_{y\in \Phi}p(y|x)log_2p(y|x)$</li>
<li>自信息：事件X发生的不确定性或事件包含的信息量$I(X)=-log_2P(x)$</li>
<li>互信息：$I(X;Y)=H(X)-H(X|Y)=log_2\frac{P(X,Y)}{P(X)P(Y)}$</li>
</ul>
</li>
</ol>
<h2>搜索系统和推荐系统常识</h2>
<p><font>ACM推荐系统大会</font></p>
<ol>
<li>搜索引擎分类：<ul>
<li>全文搜索引擎</li>
<li>元搜索引擎</li>
<li>垂直搜索引擎</li>
<li>目录搜索引擎</li>
</ul>
</li>
<li>推荐系统分类：
</li>
<li>推荐系统问题：
</li>
</ol>
<h2>知识图谱相关理论</h2>
<p>知识图谱：用图模型描述知识和建模世界万物之间关联关系的一种技术方法。组成三要素：实体、关系、属性</p>
<ol>
<li>
<strong>信息抽取</strong><ul>
<li>命名实体识别（Named Entity Recognition, NER）
</li>
<li>关系抽取
</li>
</ul>
</li>
<li>
<strong>知识融合</strong>
</li>
<li>
<strong>知识加工</strong>
</li>
<li>
<strong>质量评估</strong> </li>
</ol>
<h1>搜索系统的基本原理</h1>
<h2>搜索系统框架及原理</h2>
<ol>
<li>
<strong>数据收集及预处理</strong><ul>
<li>爬虫</li>
<li>数据清洗（去重算法<font>SimHash</font>）</li>
<li>存储空间及分布式设计（<font>Trie树</font>） </li>
</ul>
</li>
<li>
<strong>文本分析</strong><ul>
<li>查询处理<ul>
<li>分词方法：基于字符串、基于理解、基于统计的分词方法</li>
<li>查询建议 </li>
<li>查询更正（若样本类别不均衡，可采用上下采样）</li>
</ul>
</li>
<li>意图理解</li>
<li>其他方法<ul>
<li>层次聚类<ul>
<li>分类：凝聚式（自底向上）、分裂式（自顶向下）</li>
<li>求距离/相似度方法：闵可夫斯基距离（P范数）、马氏距离、相关系数、夹角余弦</li>
</ul>
</li>
<li>K均值聚类：采用欧式距离，初始聚类中心可考虑采用层次聚类</li>
<li>LDA主题模型<ul>
<li>LSA（Latent Senmantic Analysis）</li>
<li>PLSA（Probability LSA）</li>
<li>LDA（Latent Dirichlet Allocation）</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li>基于知识图谱的搜索系统</li>
</ol>
<h2>搜索系统中的主要算法</h2>
<ol>
<li>信息检索基本模型<ul>
<li>布尔模型</li>
<li>向量空间模型（权重计算：TF-IDF）</li>
<li>概率检索模型（二元独立概率模型<font>BIM</font>）</li>
<li>其他模型：基于集合论、基于代数论、基于概率统计的模型</li>
</ul>
</li>
<li>搜索和机器学习<ul>
<li>排序学习：单文档方法（CTR）、文档对方法（Boost、SVM、NN）、文档列表方法</li>
<li>示例：LR、AdaBoost、随机森林（海森矩阵：多元函数的二阶偏导构成的矩阵）</li>
</ul>
</li>
<li>搜索和深度学习<ul>
<li>DNN</li>
<li>DSSM（Deep Structured Semantic MOdels）</li>
<li>Transformer：编码、解码、Softmax与线性变换</li>
</ul>
</li>
</ol>
<h2>搜索系统评价</h2>
<ol>
<li>效率评价：响应时间和开销、索引量</li>
<li>效果评价<ul>
<li>准确率和召回率</li>
<li>平均化和插值</li>
<li>排序靠前文档质量<ul>
<li>MRR（Mean Reciprocal Rank）</li>
<li>DCG（Normalized Discounted Cumulative Gain）</li>
</ul>
</li>
</ul>
</li>
</ol>
<h1>推荐系统的基本原理</h1>
<h2>推荐系统框架及原理</h2>
<p>预测+排序+可解释性</p>
<ol>
<li>基本框架：离线层、存储层、在线层</li>
<li>经典问题<ul>
<li>探索和利用（Exploration &amp; Exploitation, EE）
</li>
<li>冷启动问题<ul>
<li>利用热门数据</li>
<li>利用用户注册信息</li>
<li>利用第三方数据</li>
<li>利用物品内容属性</li>
<li>利用专家标注数据</li>
</ul>
</li>
</ul>
</li>
<li>推荐系统的找回策略<ul>
<li>基于行为相似的召回<ul>
<li>相似度算法：Jacard相似度、余弦相似度、欧几里得相似度、皮尔逊相关系数</li>
</ul>
</li>
<li>基于内容相似的召回<ul>
<li>Hoffman编码及Hoffman Tree</li>
<li>CBow-Hierarchical Softmax</li>
<li>Skip-Gram-Hierarchical Softmax</li>
</ul>
</li>
</ul>
</li>
<li>推荐系统排序<ul>
<li>特征预处理：特征选择</li>
<li>排序模型<ul>
<li>线性模型 LR、FM、FFM</li>
<li>树模型 GDBT</li>
<li>深度学习模型</li>
<li>组合模型 GDBT+LR</li>
</ul>
</li>
</ul>
</li>
<li>基于知识图谱的推荐系统
</li>
</ol>
<h2>推荐系统的主要算法</h2>
<ol>
<li>矩阵分解<ul>
<li>奇异值分解 SVD</li>
<li>交替最小二乘 ACS，Alterating LeastSquares</li>
<li>贝叶斯个性化排序</li>
</ul>
</li>
<li>线性模型<ul>
<li>因子分解机 FM，Factorization Machine</li>
<li>FFM，Field-aware FM</li>
</ul>
</li>
<li>树模型<ul>
<li>决策树 ID3、CART、C4.5</li>
<li>集成算法 梯度提升决策树（Gradient Boosting Decision Tree, GDBT）</li>
<li>GDBT+LR</li>
<li>eXtreme Gradient Boosting, XGBoost</li>
<li>LightGBM GOSS（Gradient-based One-side Sampling）+EFB（Exclusive Feature Bunding）</li>
</ul>
</li>
<li>深度学习模型<ul>
<li>Wide &amp; Deep model
</li>
<li>Deep FM model </li>
</ul>
</li>
</ol>
<h2>推荐系统的评价</h2>
<ol>
<li>推荐系统维度<ul>
<li>RMSE（Root Mean Squre Error）+R方</li>
<li>MAP（Mean Average Precision）+ MRR（Mean Reciprocal Rank）</li>
<li>AUC + ROC</li>
</ul>
</li>
<li>离线评估：准确率、覆盖率、多样性、实时性、Robost</li>
<li>在线评估：A/B实验、Interleaving</li>
</ol>
<h1>应用</h1>
<h2>搜索引擎工具</h2>
<ol>
<li>Lucene</li>
<li>Solr</li>
<li>Elasticsearch</li>
</ol>
<h2>搜索应用实战：基于电商的搜索开发</h2>
<h2>推荐应用实战：基于广告平台的推荐</h2></article></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via 一直进步 做喜欢的 <br>
<br>
May 2, 2023 at 08:31AM
